cirq-google/cirq_google/engine/engine_client.py

```diff
@@ -95,6 +98,152 @@ def instance(cls):
+                self._request_queue = asyncio.Queue()  # Clear requests
+
+    async def _make_request(
+        self, project_name: str, program: quantum.QuantumProgram, job: quantum.QuantumJob
+    ) -> quantum.QuantumResult:
+        """This method is executed in a separate asyncio Task for each request."""
+        current_request = quantum.QuantumRunStreamRequest(
+            parent=project_name,
+            create_quantum_program_and_job=quantum.CreateQuantumProgramAndJobRequest(
+                parent=project_name, quantum_program=program, quantum_job=job
+            ),
+        )
+        get_result_request = quantum.QuantumRunStreamRequest(
+            parent=project_name, get_quantum_result=quantum.GetQuantumResultRequest(parent=job.name)
+        )
+
+        response_future: Optional[asyncio.Future] = None
+        response: Optional[quantum.QuantumRunStreamResponse] = None
+        while response is None:
+            try:
+                response_future = self._response_demux.subscribe(current_request)
+                await self._request_queue.put(current_request)
+                response = await response_future
+
+            except GoogleAPICallError:
+                # TODO how to distinguish between program not found vs job not found?
+                # TODO Send a CreateProgramAndJobRequest or CreateJobRequest if either program or
+                # job doesn't exist.
+                # TODO add exponential backoff
+                current_request = get_result_request
+                continue
+
+            # Either when this request is canceled or the _manage_stream() loop is canceled.
+            except asyncio.CancelledError:
+                if response_future is not None:
+                    response_future.cancel()
+                    self._response_demux.unsubscribe(current_request)
+                    await self._cancel(job.name)
+                    return quantum.QuantumResult()
+
+            if response.result is not None:
+                return response.result
+                # TODO handle QuantumJob response and retryable StreamError.
+
+    async def _cancel(self, job_name: str) -> None:
+        await self._grpc_client.cancel_quantum_job(quantum.CancelQuantumJobRequest(name=job_name))
+
+    def send(
+        self, project_name: str, program: quantum.QuantumProgram, job: quantum.QuantumJob
+    ) -> duet.AwaitableFuture[quantum.QuantumResult]:
+        """Sends a request over the stream and returns a future for the result."""
+        if self._manage_stream_loop_future is None:
+            self._manage_stream_loop_future = self._executor.submit(self._manage_stream)
+        return self._executor.submit(self._make_request, project_name, program, job)
+
+    def stop(self) -> None:
+        """Stops and resets the stream manager."""
+        if self._manage_stream_loop_future is not None:
+            self._manage_stream_loop_future.cancel()
+            self._manage_stream_loop_future = None
```

Comment:
I'd suggest returning a future here too, so the caller can await to ensure that cancellation is done:
```suggestion
    def stop(self) -> duet.AwaitableFuture[None]:
        """Stops and resets the stream manager."""
        if self._manage_stream_loop_future is None:
            return duet.completed_future(None)
        self._manage_stream_loop_future.cancel()
        return self._manage_stream_loop_future
```
You might need to do some work in `_manage_future` because currently it looks like it will run forever even if a cancelation error is raised internally since it captures `BaseException`, though it might be that asyncio will just stop running the coroutine when a task is cancelled.

Reply:
From my experiments, `cancel()` waits for the done callback to finish, but it doesn't wait for the `CancelledError` handler to finish. Awaiting on the future doesn't wait for `CancelledError` handler, either.

Let me know if there's a better way to wait for the error handler other than artificially introducing a delay. I'll revert the change for the time being.